### DISCUSSION TITLE
“Material has no property named: dfundgrad_op”

### DISCUSSION THREAD

**Discussion by dongcongcong606** [idaholab/moose#18446] [Q&A Modules: Phase field] [answered]
Hi everyone!
I am new to Moose. And I have declared  "dfundgrad_op" in Materials. But when I executed my code, there was still an error suggesting that "Material has no property named: dfundgrad_op”.  I was confused, but I don't know how to deal with it. May someone help me? Thank you very much! 

...H

 MaterialProperty<RealGradient> & _dfundgrad_op;
.....

...C

    _dfundgrad_op(declareProperty<RealGradient>("dfundgrad_op")),
....

**Comment by hugary1995** [comment, marked as answer]
Does your code compile without error? The first obvious mistake is `MaterialProperty & _dfundgrad_op;`, which should be specialized to `MaterialProperty<RealGradient> & _dfundgrad_op;`.

A general suggestion: if you are new to moose, I highly recommend you to go through the workshop. It's a bit lengthy, but definitely worth your time.

**Reply by dongcongcong606** [reply]
Thank you, first. My code may be grammatically correct. (no error was reported). However, when I execute the input file, an error occurred.  By the way,  what does "specialize" mean? I just choose a "private" for it as follows. Thank you very much!

material.h(for Material)
  xxxxx
private:

  xxxxxx
  MaterialProperty<RealGradient> & _dfundgrad_op;
.........................................................................................................................................

SplitCHWBase3.h(for kernel)

template <typename T>
class SplitCHWResBase3 : public DerivativeMaterialInterface<JvarMapKernelInterface<Kernel>>
{
public:
  static InputParameters validParams();

  SplitCHWResBase3(const InputParameters & parameters);

protected:

      xxxxxx

  const MaterialProperty<T> & _dfundgrad_op;
};
............................................................................................................................................

**Reply by hugary1995** [reply]
I'm puzzled as to why it even compiles... Let's see if someone else can help.

**Reply by GiudGiud** [reply]
Maybe those files are not being compiled, like the MOOSE_DIR is set somewhere else. 
You should get this error
```
In file included from /home/guillaume/projects/moose/framework/src/materials/GenericConstantMaterial.C:10,
                 from /home/guillaume/projects/moose/framework/build/unity_src/materials_Unity.C:12:
/home/guillaume/projects/moose/framework/build/header_symlinks/GenericConstantMaterial.h:38:3: error: template placeholder type 'MaterialProperty<...auto...>' must be followed by a simple declarator-id
   38 |   MaterialProperty & _dfundgrad_op;
      |   ^~~~~~~~~~~~~~~~
```

specialize means that you give the templated class an argument for the template.
For example vector<T> is a template for vector. If I want a vector of Real, you declare the object as vector<Real>. The compiler will know to place Real everywhere in the definition of vector<T> (for example the routines to say how an addition is done, how to take the first element etc)

private / public is about access management, see [here](https://www.tutorialspoint.com/cplusplus/cpp_class_access_modifiers.htm)

**Reply by dongcongcong606** [reply]
Thank you very much ， I have got it.  The two types did not match when I converted them, but the compiler did not report the error, and it took me a long time to find the problem. Thank you for your tips and suggestions again.
